Backend/Remora.Discord.Gateway/Services/DefaultResponderDispatchService.cs

```diff
@@ -0,0 +1,391 @@
+    private readonly Channel<Task<IReadOnlyList<Result>>> _responderFinalizationQueue;
+    private readonly Type _dispatcherType;
+    private readonly Dictionary<Type, Type> _eventTypeInterfaces;
+    private readonly Dictionary<Type, MethodInfo> _dispatchMethods;
+
+    private CancellationTokenSource _dispatchCts;
+    private Task<Result>? _finalizerTask;
+    private bool _isDisposed;
+
+    /// <inheritdoc />
+    public bool IsRunning { get; private set; }
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="DefaultResponderDispatchService"/> class.
+    /// </summary>
+    /// <param name="logger">The logging provider.</param>
+    /// <param name="responderTypeRepository">The payload handler type repository.</param>
+    /// <param name="services">The service provider.</param>
+    public DefaultResponderDispatchService
+    (
+        ILogger<DefaultResponderDispatchService> logger,
+        IResponderTypeRepository responderTypeRepository,
+        IServiceProvider services
+    )
+    {
+        _logger = logger;
+        _responderTypeRepository = responderTypeRepository;
+        _services = services;
+
+        _eventDispatchQueue = Channel.CreateUnbounded<IGatewayEvent>();
+        _responderFinalizationQueue = Channel.CreateUnbounded<Task<IReadOnlyList<Result>>>();
+        _dispatcherType = GetType();
+        _eventTypeInterfaces = new Dictionary<Type, Type>();
+        _dispatchMethods = new Dictionary<Type, MethodInfo>();
+
+        _dispatchCts = new CancellationTokenSource();
+    }
+
+    /// <inheritdoc />
+    public async ValueTask EnqueueEventAsync<TGatewayEvent>(TGatewayEvent gatewayEvent, CancellationToken ct = default)
+        where TGatewayEvent : IGatewayEvent
+    {
+        await _eventDispatchQueue.Writer.WriteAsync(gatewayEvent, ct);
+    }
+
+    /// <inheritdoc />
+    public async Task<Result> RunAsync(CancellationToken ct = default)
+    {
+        if (this.IsRunning)
+        {
+            return new InvalidOperationError("This dispatch service is already running");
+        }
+
+        try
+        {
+            await Task.Yield();
+
+            // Reset our state
+            _dispatchCts.Dispose();
+            _dispatchCts = new CancellationTokenSource();
+
+            _finalizerTask?.Dispose();
+            _finalizerTask = FinalizerTaskAsync();
+
+            this.IsRunning = true;
+
+            try
+            {
+                await foreach (var gatewayEvent in _eventDispatchQueue.Reader.ReadAllAsync(ct))
+                {
+                    var dispatchResult = DispatchEvent(gatewayEvent);
+                    if (!dispatchResult.IsSuccess)
+                    {
+                        return Result.FromError(dispatchResult);
+                    }
+
+                    await _responderFinalizationQueue.Writer.WriteAsync(dispatchResult.Entity, ct);
+
+                    if (_finalizerTask.IsCompleted)
```

Comment:
Is this meant to be inverted? `awaiting` a completed task is essentially a no-op, and given that it's completed, you could return `.Result`, though at a glance you may get a few looks for that.

Reply:
I agree; it should be inverted so that this check is made before the rest of the logic. At least, I assume that's what you meant?

As to `awaiting` it, am I mistaken in thinking that it will resolve to returning the result directly? I see no harm in changing it though. I've kept consistent with how Jax has written similar logic - see [here](https://github.com/Nihlus/Remora.Discord/blob/feaf50c481cbdbdab025635aa73804a7f7a4ed8f/Backend/Remora.Discord.Gateway/DiscordGatewayClient.cs#L591).

Reply:
Hmm, looks like Github selected more than intended; I was referring to the boolean check itself, hence the comment on awaiting a completed task. Consistency is good, though.

Reply:
Ah, gotcha. No, the finalizer task should run for the lifetime of the dispatch service. If it has completed then something has gone wrong, so I return that error. It is then up to the consumer to inspect the error and restart the dispatch service.